pkg/controllers/routing/network_routes_controller.go

```diff
@@ -508,130 +506,174 @@ func (nrc *NetworkRoutingController) advertisePodRoute() error {
-			}
-			if err := netlink.LinkSetUp(link); err != nil {
-				return errors.New("Failed to bring tunnel interface " + tunnelName + " up due to: " + err.Error())
-			}
-			// reduce the MTU by 20 bytes to accommodate ipip tunnel overhead
-			if err := netlink.LinkSetMTU(link, link.Attrs().MTU-20); err != nil {
-				return errors.New("Failed to set MTU of tunnel interface " + tunnelName + " up due to: " + err.Error())
-			}
-		} else {
-			klog.Infof("Tunnel interface: " + tunnelName + " for the node " + nextHop.String() + " already exists.")
-		}
-
-		out, err := exec.Command("ip", "route", "list", "table", customRouteTableID).CombinedOutput()
-		if err != nil || !strings.Contains(string(out), "dev "+tunnelName+" scope") {
-			if out, err = exec.Command("ip", "route", "add", nextHop.String(), "dev", tunnelName, "table",
-				customRouteTableID).CombinedOutput(); err != nil {
-				return fmt.Errorf("failed to add route in custom route table, err: %s, output: %s", err, string(out))
-			}
+			return err
 		}
+	} else {
+		// knowing that a tunnel shouldn't exist for this route, check to see if there are any lingering tunnels /
+		// routes that need to be cleaned up.
+		nrc.cleanupTunnel(dst, tunnelName)
+	}
 
+	if link != nil {
+		// if we setup an overlay tunnel link, then use it for destination routing
 		route = &netlink.Route{
 			LinkIndex: link.Attrs().Index,
 			Src:       nrc.nodeIP,
 			Dst:       dst,
 			Protocol:  0x11,
 		}
 	} else if sameSubnet {
+		// if the nextHop is within the same subnet, add a route for the destination so that traffic can bet routed
+		// at layer 2 and minimize the need to traverse a router
 		route = &netlink.Route{
 			Dst:      dst,
 			Gw:       nextHop,
 			Protocol: 0x11,
 		}
 	} else {
+		// otherwise, let BGP do its thing, nothing to do here
 		return nil
 	}
 
-	if path.IsWithdraw {
-		klog.V(2).Infof("Removing route: '%s via %s' from peer in the routing table", dst, nextHop)
-		return netlink.RouteDel(route)
-	}
+	// Alright, everything is in place, and we have our route configured, let's add it to the host's routing table
 	klog.V(2).Infof("Inject route: '%s via %s' from peer to routing table", dst, nextHop)
 	return netlink.RouteReplace(route)
 }
 
+func (nrc *NetworkRoutingController) isPeerEstablished(peerIP string) (bool, error) {
+	var peerDisconnected bool
+	peerFunc := func(peer *gobgpapi.Peer) {
+		if peer.Conf.NeighborAddress == peerIP {
+			if peer.State.SessionState != gobgpapi.PeerState_ESTABLISHED {
+				peerDisconnected = true
+			}
+		}
+	}
+	err := nrc.bgpServer.ListPeer(context.Background(), &gobgpapi.ListPeerRequest{
+		Address: peerIP,
+	}, peerFunc)
+	if err != nil {
+		return false, fmt.Errorf("unable to list peers to see if tunnel & routes need to be removed: %v", err)
+	}
+
+	return peerDisconnected, nil
+}
+
+// cleanupTunnels removes any traces of tunnels / routes that were setup by nrc.setupOverlayTunnel() and are no longer
```

Comment:
this should be cleanupTunnel (singular), right?

Reply:
Sure enough. Fixed.